Crashes on Linux and Windows

**Describe the bug**
I tried the latest 2.4.3 on both Windows and Linux (latest version of Mint), and both crash in the browser as I am trying to authenticate.
I also tried 2.4.2 on Linux, but Slack will not allow that browser to log in (too old).

**Output**
Crash backtrace:
```./slackdump -u
Slackdump 2.4.3 (commit: ef85584f863e6cb20c036aa672394fbdadc8b88f) built on: 2023-10-06T23:14:55Z
? Enter Slack Workspace Name:  xxx
Please login in the browser...
BEWARE: your OS is not officially supported by Playwright; downloading fallback build.
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x82ba3c]

goroutine 1 [running]:
github.com/rusq/slackdump/v2/auth/browser.extractToken({0x0, 0x0})
	/home/runner/work/slackdump/slackdump/auth/browser/extractors.go:29 +0x1c
github.com/rusq/slackdump/v2/auth/browser.(*Client).Authenticate(0xc0000220c0, {0xa74248?, 0xc0001487e0?})
	/home/runner/work/slackdump/slackdump/auth/browser/client.go:119 +0x911
github.com/rusq/slackdump/v2/auth.NewBrowserAuth({0xa74248, 0xc0001487e0}, {0xc0001292e8, 0x2, 0xc0000061a0?})
	/home/runner/work/slackdump/slackdump/auth/browser.go:62 +0x365
github.com/rusq/slackdump/v2/internal/app.SlackCreds.AuthProvider({{0x0?, 0x8f1ca0?}, {0x0?, 0x7b0052?}}, {0xa74248, 0xc0001487e0}, {0x0, 0x0}, 0x0)
	/home/runner/work/slackdump/slackdump/internal/app/auth.go:77 +0x185
github.com/rusq/slackdump/v2/internal/app.InitProvider({0xa74248?, 0xc0001487b0?}, {0xc000104360, 0x1f}, {0x0, 0x0}, {0xa72580, 0xc0000ba700}, 0xc0000ade60?)
	/home/runner/work/slackdump/slackdump/internal/app/auth.go:149 +0x310
main.run({_, _}, {{{0x1, 0x0}, {0xc0000ade60}, {{0xa6c398, 0x1}, {0x98fdde, 0x4}, {0x0, ...}}, ...}, ...})
	/home/runner/work/slackdump/slackdump/cmd/slackdump/main.go:135 +0x20d
main.main()
	/home/runner/work/slackdump/slackdump/cmd/slackdump/main.go:106 +0x566
```

**Desktop (please complete the following information):**
 - OS: Linux Mint or Windows-10

--- a/auth/browser.go
+++ b/auth/browser.go
@@ -6,6 +6,7 @@ import (
 	"net/url"
 	"os"
 	"strings"
+	"time"
 
 	"github.com/rusq/slackdump/v2/auth/auth_ui"
 	"github.com/rusq/slackdump/v2/auth/browser"
@@ -20,9 +21,10 @@ type BrowserAuth struct {
 }
 
 type browserOpts struct {
-	workspace string
-	browser   browser.Browser
-	flow      BrowserAuthUI
+	workspace    string
+	browser      browser.Browser
+	flow         BrowserAuthUI
+	loginTimeout time.Duration
 }
 
 type BrowserAuthUI interface {
@@ -33,8 +35,9 @@ type BrowserAuthUI interface {
 func NewBrowserAuth(ctx context.Context, opts ...Option) (BrowserAuth, error) {
 	var br = BrowserAuth{
 		opts: browserOpts{
-			flow:    defaultFlow,
-			browser: browser.Bfirefox,
+			flow:         defaultFlow,
+			browser:      browser.Bfirefox,
+			loginTimeout: browser.DefLoginTimeout,
 		},
 	}
 	for _, opt := range opts {
@@ -55,7 +58,7 @@ func NewBrowserAuth(ctx context.Context, opts ...Option) (BrowserAuth, error) {
 		br.opts.workspace = wsp
 	}
 
-	auther, err := browser.New(br.opts.workspace, browser.OptBrowser(br.opts.browser))
+	auther, err := browser.New(br.opts.workspace, browser.OptBrowser(br.opts.browser), browser.OptTimeout(br.opts.loginTimeout))
 	if err != nil {
 		return br, err
 	}

--- a/auth/browser/browser.go
+++ b/auth/browser/browser.go
@@ -3,9 +3,12 @@ package browser
 import (
 	"fmt"
 	"strings"
+	"time"
 
 	"github.com/playwright-community/playwright-go"
 )
+
+const defLoginTimeout = 5 * time.Minute // default login timeout.
 
 //go:generate stringer -type Browser -trimprefix=B browser.go
 type Browser int
@@ -23,6 +26,15 @@ func OptBrowser(b Browser) Option {
 			b = Bfirefox
 		}
 		c.br = b
+	}
+}
+
+func OptTimeout(d time.Duration) Option {
+	return func(c *Client) {
+		if d < 0 {
+			return
+		}
+		c.loginTimeout = float64(d.Milliseconds())
 	}
 }
 

--- a/auth/browser/browser.go
+++ b/auth/browser/browser.go
@@ -8,7 +8,8 @@ import (
 	"github.com/playwright-community/playwright-go"
 )
 
-const defLoginTimeout = 5 * time.Minute // default login timeout.
+// DefLoginTimeout is the default Slack login timeout
+const DefLoginTimeout = 5 * time.Minute
 
 //go:generate stringer -type Browser -trimprefix=B browser.go
 type Browser int

--- a/auth/browser/client.go
+++ b/auth/browser/client.go
@@ -125,9 +125,6 @@ func (cl *Client) Authenticate(ctx context.Context) (string, []*http.Cookie, err
 	}); err != nil {
 		return "", nil, err
 	}
-	if r == nil {
-		return "", nil, errors.New("failed waiting for the request")
-	}
 
 	token, err := extractToken(r)
 	if err != nil {

--- a/auth/browser/client.go
+++ b/auth/browser/client.go
@@ -115,6 +115,9 @@ func (cl *Client) Authenticate(ctx context.Context) (string, []*http.Cookie, err
 	}); err != nil {
 		return "", nil, err
 	}
+	if r == nil {
+		return "", nil, errors.New("failed waiting for the request")
+	}
 
 	token, err := extractToken(r)
 	if err != nil {

--- a/auth/browser/client.go
+++ b/auth/browser/client.go
@@ -39,7 +39,7 @@ func New(workspace string, opts ...Option) (*Client, error) {
 		workspace:    workspace,
 		pageClosed:   make(chan bool, 1),
 		br:           Bfirefox,
-		loginTimeout: float64(defLoginTimeout.Milliseconds()),
+		loginTimeout: float64(DefLoginTimeout.Milliseconds()),
 	}
 	for _, opt := range opts {
 		opt(cl)

--- a/auth/browser/client.go
+++ b/auth/browser/client.go
@@ -20,9 +20,10 @@ const slackDomain = ".slack.com"
 
 // Client is the client for Browser Auth Provider.
 type Client struct {
-	workspace  string
-	pageClosed chan bool // will receive a notification that the page is closed prematurely.
-	br         Browser
+	workspace    string
+	pageClosed   chan bool // will receive a notification that the page is closed prematurely.
+	br           Browser
+	loginTimeout float64 // slack login page timeout in milliseconds.
 }
 
 var Logger logger.Interface = logger.Default
@@ -34,7 +35,12 @@ func New(workspace string, opts ...Option) (*Client, error) {
 	if workspace == "" {
 		return nil, errors.New("workspace can't be empty")
 	}
-	cl := &Client{workspace: workspace, pageClosed: make(chan bool, 1), br: Bfirefox}
+	cl := &Client{
+		workspace:    workspace,
+		pageClosed:   make(chan bool, 1),
+		br:           Bfirefox,
+		loginTimeout: float64(defLoginTimeout.Milliseconds()),
+	}
 	for _, opt := range opts {
 		opt(cl)
 	}
@@ -100,6 +106,7 @@ func (cl *Client) Authenticate(ctx context.Context) (string, []*http.Cookie, err
 	if err != nil {
 		return "", nil, err
 	}
+	// page close sentinel.
 	page.On("close", func() { trace.Log(ctx, "user", "page closed"); close(cl.pageClosed) })
 
 	uri := fmt.Sprintf("https://%s"+slackDomain, cl.workspace)
@@ -110,8 +117,11 @@ func (cl *Client) Authenticate(ctx context.Context) (string, []*http.Cookie, err
 	}
 
 	var r playwright.Request
-	if err := cl.withBrowserGuard(ctx, func() {
-		r, err = page.ExpectRequest(uri+"/api/api.features*", func() error { return nil })
+	if err := cl.withBrowserGuard(ctx, func() error {
+		r, err = page.ExpectRequest(uri+"/api/api.features*", func() error { return nil }, playwright.PageExpectRequestOptions{
+			Timeout: &cl.loginTimeout,
+		})
+		return err
 	}); err != nil {
 		return "", nil, err
 	}
@@ -135,20 +145,27 @@ func (cl *Client) Authenticate(ctx context.Context) (string, []*http.Cookie, err
 	return token, convertCookies(state.Cookies), nil
 }
 
-func (cl *Client) withBrowserGuard(ctx context.Context, fn func()) error {
-	done := make(chan struct{})
+var errPageClosed = errors.New("browser closed or timed out")
+
+// withBrowserGuard starts the function fn in a goroutine, and waits for it to
+// finish.  If the context is canceled, or the page is closed, it returns
+// the appropriate error.
+func (cl *Client) withBrowserGuard(ctx context.Context, fn func() error) error {
+	var (
+		errC = make(chan error)
+	)
 	go func() {
-		defer close(done)
-		fn()
+		defer close(errC)
+		errC <- fn()
 	}()
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	case <-cl.pageClosed:
-		return errors.New("browser closed or timed out")
-	case <-done:
+		return errPageClosed
+	case err := <-errC:
+		return err
 	}
-	return nil
 }
 
 func convertCookies(pwc []playwright.Cookie) []*http.Cookie {

--- a/auth/browser/extractors.go
+++ b/auth/browser/extractors.go
@@ -26,6 +26,9 @@ var (
 
 // extractToken extracts token from the request.
 func extractToken(r playwright.Request) (string, error) {
+	if r == nil {
+		return "", errors.New("no request")
+	}
 	if r.Method() == http.MethodGet {
 		return extractTokenGet(r.URL())
 	} else if r.Method() == http.MethodPost {

--- a/auth/option.go
+++ b/auth/option.go
@@ -1,6 +1,10 @@
 package auth
 
-import "github.com/rusq/slackdump/v2/auth/browser"
+import (
+	"time"
+
+	"github.com/rusq/slackdump/v2/auth/browser"
+)
 
 type options struct {
 	*browserOpts
@@ -26,5 +30,14 @@ func BrowserWithWorkspace(name string) Option {
 func BrowserWithBrowser(b browser.Browser) Option {
 	return func(o *options) {
 		o.browserOpts.browser = b
+	}
+}
+
+func BrowserWithTimeout(d time.Duration) Option {
+	return func(o *options) {
+		if d < 0 {
+			return
+		}
+		o.browserOpts.loginTimeout = d
 	}
 }

--- a/cmd/slackdump/main.go
+++ b/cmd/slackdump/main.go
@@ -11,6 +11,7 @@ import (
 	"path/filepath"
 	"runtime/trace"
 	"syscall"
+	"time"
 
 	"github.com/joho/godotenv"
 	"github.com/rusq/dlog"
@@ -52,10 +53,11 @@ var secrets = []string{".env", ".env.txt", "secrets.txt"}
 
 // params is the command line parameters
 type params struct {
-	appCfg    config.Params
-	creds     app.SlackCreds
-	authReset bool
-	browser   browser.Browser
+	appCfg         config.Params
+	creds          app.SlackCreds
+	authReset      bool
+	browser        browser.Browser
+	browserTimeout time.Duration
 
 	traceFile string // trace file
 	logFile   string //log file, if not specified, outputs to stderr.
@@ -253,6 +255,7 @@ func parseCmdLine(args []string) (params, error) {
 	fs.StringVar(&p.creds.Cookie, "cookie", osenv.Secret(envSlackCookie, ""), "d= cookie `value` or a path to a cookie.txt file (environment: "+envSlackCookie+")")
 	fs.BoolVar(&p.authReset, "auth-reset", false, "reset EZ-Login 3000 authentication.")
 	fs.Var(&p.browser, "browser", "set the browser to use for authentication: 'chromium' or 'firefox' (default: firefox)")
+	fs.DurationVar(&p.browserTimeout, "browser-timeout", browser.DefLoginTimeout, "browser login timeout")
 	fs.StringVar(&p.workspace, "w", "", "set the Slack `workspace` name.  If not specifed, the slackdump will show an\ninteractive prompt.")
 
 	// operation mode

--- a/cmd/slackdump/main_test.go
+++ b/cmd/slackdump/main_test.go
@@ -8,6 +8,7 @@ import (
 	"github.com/stretchr/testify/assert"
 
 	"github.com/rusq/slackdump/v2"
+	"github.com/rusq/slackdump/v2/auth/browser"
 	"github.com/rusq/slackdump/v2/internal/app"
 	"github.com/rusq/slackdump/v2/internal/app/config"
 	"github.com/rusq/slackdump/v2/internal/structures"
@@ -63,15 +64,17 @@ func Test_checkParameters(t *testing.T) {
 					Token:  "x",
 					Cookie: "d",
 				},
+				browserTimeout: browser.DefLoginTimeout,
 				appCfg: config.Params{
 					ListFlags: config.ListFlags{
 						Users:    false,
 						Channels: true,
 					},
 					FilenameTemplate: defFilenameTemplate,
-					Input:            config.Input{List: &structures.EntityList{}},
-					Output:           config.Output{Filename: "-", Format: "text"},
-					Options:          slackdump.DefOptions,
+
+					Input:   config.Input{List: &structures.EntityList{}},
+					Output:  config.Output{Filename: "-", Format: "text"},
+					Options: slackdump.DefOptions,
 				}},
 			false,
 		},
@@ -83,6 +86,7 @@ func Test_checkParameters(t *testing.T) {
 					Token:  "x",
 					Cookie: "d",
 				},
+				browserTimeout: browser.DefLoginTimeout,
 				appCfg: config.Params{
 					ListFlags: config.ListFlags{
 						Channels: false,

